wicker/core/datasets.py

```diff
@@ -125,79 +135,134 @@ def __len__(self) -> int:
         pass
 
     @abc.abstractmethod
-    def schema(self) -> DatasetSchema:
-        """Return the schema of the dataset."""
+    def arrow_table(self) -> pyarrow.Table:
+        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
         pass
 
     @abc.abstractmethod
-    def arrow_table(self) -> pyarrow.Table:
-        """Return the pyarrow table with all the metadata fields and pointers of the dataset."""
+    def schema(self) -> DatasetSchema:
+        """Return the schema of the dataset."""
         pass
 
 
-class S3Dataset(AbstractDataset):
-    """Implementation for a Map-based dataset"""
+class BaseDataset(AbstractDataset):
```

Comment:
Do we want to use inherit  or combination to implement the Abstract interface. To me, it seems we lean to the combination style, since we inject the dependency. If so, prob BaseDataset->DatasetImpl and prob the Abstract interface is not necessary, the external customer would not expect to use the arrow directly. WDYT?

Reply:
Let me mark as resolved since we are adding a factory class that users can use to create instances instead, so we may not have to worry so much about the `BaseDataset` class. But please feel free to reopen and let me know your thoughts about any changes we should make to this class.